enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/slotted/pipes/TopSlottedPipe.scala

```diff
@@ -0,0 +1,157 @@
+      val aVal = a.getRefAt(columnSlot)
+      val bVal = b.getRefAt(columnSlot)
+      order.compareValues(aVal, bVal)
+    }
+  }
+
+  def binarySearch(array: Array[ExecutionContext], comparator: Comparator[ExecutionContext])(key: ExecutionContext) = {
+    java.util.Arrays.binarySearch(array.asInstanceOf[Array[ExecutionContext]], key, comparator)
+  }
+}
+
+case class TopNSlottedPipe(source: Pipe, orderBy: Seq[ColumnOrder], countExpression: Expression)
+                           (val id: Id = new Id) extends TopSlottedPipe(source, orderBy) {
+
+  countExpression.registerOwningPipe(this)
+
+  protected override def internalCreateResults(input:Iterator[ExecutionContext], state: QueryState): Iterator[ExecutionContext] = {
+    implicit val s = state
+    if (input.isEmpty)
+      Iterator.empty
+    else if (orderBy.isEmpty)
+      input
+    else {
+      val first = input.next()
+      val count = countExpression(first).asInstanceOf[NumberValue].longValue().toInt
+      val topTable = new DefaultComparatorTopTable(comparator, count)
+      topTable.add(first)
+
+      input.foreach {
+        ctx =>
+          topTable.add(ctx)
+      }
+
+      topTable.sort()
+
+      topTable.iterator.asScala
+    }
+  }
+}
+
+/*
+ * Special case for when we only have one element, in this case it is no idea to store
+ * an array, instead just store a single value.
+ */
+case class Top1SlottedPipe(source: Pipe, orderBy: List[ColumnOrder])
+                          (val id: Id = new Id)
+  extends TopSlottedPipe(source, orderBy) {
+
+  protected override def internalCreateResults(input: Iterator[ExecutionContext],
+                                               state: QueryState): Iterator[ExecutionContext] = {
+    implicit val s = state
+    if (input.isEmpty)
+      Iterator.empty
+    else if (orderBy.isEmpty)
+      input
+    else {
+
+      val first = input.next()
+      var result = first
+
+      input.foreach {
+        ctx =>
+          if (comparator.compare(ctx, result) < 0) {
+            result = ctx
+          }
+      }
+      Iterator.single(result)
+    }
+  }
+}
+
+/*
+ * Special case for when we only want one element, and all others that have the same value (tied for first place)
+ */
+case class Top1WithTiesSlottedPipe(source: Pipe, orderBy: List[ColumnOrder])
+                                  (val id: Id = new Id)
+  extends TopSlottedPipe(source, orderBy) {
+
+  protected override def internalCreateResults(input: Iterator[ExecutionContext],
+                                               state: QueryState): Iterator[ExecutionContext] = {
+    implicit val s = state
+    if (input.isEmpty)
+      Iterator.empty
+    else {
+      val first = input.next()
+      var best = first
+      var matchingRows = init(best)
+
+      input.foreach {
+        ctx =>
+          val comparison = comparator.compare(ctx, best)
+          if (comparison < 0) { // Found a new best
+            best = ctx
+            matchingRows = init(ctx)
```

Comment:
Presumably this could happen very often. 
Can we do `matchingRows.clear()` instead?
(although `VectorBuilder` does very similar things in constructor as it does in `clear()` at least we'd create fewer `VectorBuilder` instances)

Reply:
Sure